### PR TITLE
Add support for private esc function names

### PR DESCRIPTION
**Pull request by mweagle**
Add a `-private` command option to generate package private function names.  Standard `esc` exported functions will be prefixed by _esc_.

Default behavior remains package exported.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -26,6 +26,8 @@ The flags are:
 	regular expression for files to ignore
 -modtime=""
 	Unix timestamp to override as modification time for all files
+-private
+	Use non-exported function names. Standard FS* functions will be prefixed with esc.  Eg: FSString => escFSString.
```
Comment (mjibson) [inline README.md]: Use the same style as the other lines:

- no period at EOL or in the line, use a semicolon if needed
- no caps at start except for proper names

Maybe:

`unexport functions by prefixing them with esc; e.g., FS -> escFS`

File: main.go [hunk 2]
```diff
@@ -23,9 +24,20 @@ var (
 	flagPrefix  = flag.String("prefix", "", "Prefix to strip from filesnames.")
 	flagIgnore  = flag.String("ignore", "", "Regexp for files we should ignore (for example \\\\.DS_Store).")
 	flagModTime = flag.String("modtime", "", "Unix timestamp to override as modification time for all files.")
+	flagPrivate = flag.Bool("private", false, "If true, do not export autogenerated functions")
```
Comment (mjibson) [inline main.go]: period at EOL

File: main.go [hunk 3]
```diff
@@ -156,8 +172,37 @@ func segment(s *bytes.Buffer) string {
...
 }
 
+func templateFunctionName(basename string, isPublic bool) string {
+	if isPublic {
+		return basename
+	}
+	return fmt.Sprintf("esc%s", basename)
+}
+
+func header(packageName string, enableExports bool) (string, error) {
```
Comment (mjibson) [inline main.go]: No opposition to using the template package, but I don't think it needs all of this. It should just have the minimum required fields: package name, function prefix (i.e., `esc` or empty string). No need for each thing to have its own property.

File: main.go (hunk 2 above)
Comment (mweagle) [inline main.go]: Good catch :)

File: main.go (hunk 3 above)
Comment (mweagle) [inline main.go]: Ok - will refactor.  Thanks.

**Comment by mweagle** [comment]
Hi @mjibson - checking back to see if you'd like additional changes to the PR.  Thanks.

File: main.go [hunk 4]
```diff
@@ -288,27 +332,27 @@ func (f *_escFile) Sys() interface{} {
 	return f
 }
 
-// FS returns a http.Filesystem for the embedded assets. If useLocal is true,
+// {{.FunctionPrefix}}{{.FSFunctionName}} returns a http.Filesystem for the embedded assets. If useLocal is true,
```
Comment (mjibson) [inline main.go]: I'd prefer to hard code FSFunctionName, DirFunctionName, etc. Only leave as parameters things that can change (currently just FunctionPrefix and PackageName).

**Comment by mjibson** [comment]
Github doesn't email after new commits are pushed. Post a comment after you push new commits and I'll look.

**Comment by mweagle** [comment]
Will do - thanks.

**Comment by mweagle** [comment]
@mjibson - Updated to limit parameter expansion to only varying values.

File: main.go [hunk 5]
```diff
@@ -156,8 +167,29 @@ func segment(s *bytes.Buffer) string {
 	return res + "`"
 }
 
+func header(packageName string, enableExports bool) (string, error) {
+	functionPrefix := ""
+	if !enableExports {
+		functionPrefix = "esc"
```
Comment (mjibson) [inline main.go]: `_esc` would be more consistent with the rest of the names.

File: main.go [hunk 6]
```diff
@@ -156,8 +167,29 @@ func segment(s *bytes.Buffer) string {
 	return res + "`"
 }
 
+func header(packageName string, enableExports bool) (string, error) {
```
Comment (mjibson) [inline main.go]: return `([]byte, error)` instead

File: main.go [hunk 7]
```diff
@@ -98,7 +105,11 @@ func main() {
 		}
 		defer w.Close()
 	}
-	fmt.Fprintf(w, header, *flagPkg)
+	headerText, err := header(*flagPkg, !(*flagPrivate))
+	if nil != err {
+		log.Fatalf("Failed to expand autogenerated code: %v", err)
+	}
+	fmt.Fprint(w, headerText)
```
Comment (mjibson) [inline main.go]: With the `[]byte` return below, you can do just `w.Write(headerText)`.

And I know it wasn't done before, but could you add error checking to that as well?

```
if _, err := w.Write(headerText); err != nil {
  log.Fatalf("failed to write output: %s", err)
}
```

File: main.go [hunk 8]
```diff
@@ -98,7 +105,11 @@ func main() {
 		}
 		defer w.Close()
 	}
-	fmt.Fprintf(w, header, *flagPkg)
+	headerText, err := header(*flagPkg, !(*flagPrivate))
+	if nil != err {
+		log.Fatalf("Failed to expand autogenerated code: %v", err)
```
Comment (mjibson) [inline main.go]: Failed -> failed
%v -> %s

**Comment by mweagle** [comment]
Thx for feedback - PR updated with latest comments.

**Comment by mweagle** [comment]
The leading underscore identifiers collide with existing types.  Closing & will reopen.

**Pull request closed without merging**